pkg/identity/cache.go

```diff
@@ -119,24 +119,39 @@ func LookupIdentity(lbls labels.Labels) *Identity {
 // LookupReservedIdentityByLabels looks up a reserved identity by its labels and
 // returns it if found. Returns nil if not found.
 func LookupReservedIdentityByLabels(lbls labels.Labels) *Identity {
-	// If there is only one label with the "reserved" source and a well-known
-	// key, return the well-known identity for that key.
-	if len(lbls) != 1 {
-		return nil
-	}
+	var (
+		idty                *Identity
+		containsNonReserved bool
+	)
 	for _, lbl := range lbls {
-		if lbl.Source != labels.LabelSourceReserved {
-			return nil
-		}
-		if id := GetReservedID(lbl.Key); id != IdentityUnknown {
-			return LookupReservedIdentity(id)
+		switch {
+		// If the set of labels contain a fixed identity then and exists in
+		// the map of reserved IDs then return the identity of that reserved ID.
+		case lbl.Key == labels.LabelKeyFixedIdentity:
+			id := GetReservedID(lbl.Value)
+			if id != IdentityUnknown && IsUserReservedIdentity(id) {
+				return LookupReservedIdentity(id)
+			}
+		// If it doesn't contain a fixed-identity then make sure the set of
+		// labels only contains a single label and that label is of the reserved
+		// type. This is to prevent users from adding cilium-reserved labels
+		// into the workloads.
+		case lbl.Source != labels.LabelSourceReserved:
```

Comment:
I would change this completely and instead do:

```
case lbl.Source == labels.LabelSourceReserved:
     if len(lbls) != 1 {
        return nil
    }
    id := GetReservedID(lbl.Key)
    if id != IdentityUnknown && !IsUserReservedIdentity(id) {
        return id
    }
```

Reply:
Done